### ISSUE TITLE
409 error when try to get skype staticfiles (emotions)

### ISSUE THREAD

**Issue by mihalt** [Terrance/SkPy#268]
### Before we start...

- [X] I've searched existing issues, but my problem hasn't been reported yet.
- [X] I've read the documentation (including notes on error messages and rate limiting), but my problem is something else.
- [X] I've tested the behaviour on [Skype for Web](https://web.skype.com), and it works there but not with SkPy.

### Summary

When I call SkypeUtils.static["items"], I get SkypeApiException with 409 error

That's how the link is looks like in browser from loggined account
![изображение](https://github.com/user-attachments/assets/2aa5f11a-959c-4d60-8118-de1799b4f7cd)

### Code sample

```python
SkypeUtils.static["items"]
```

### Code output

```python
Traceback (most recent call last):
2024-08-10T12:42:28.759916343Z   File "/code/manage.py", line 24, in <module>
2024-08-10T12:42:28.762766161Z     main()
2024-08-10T12:42:28.762796662Z   File "/code/manage.py", line 20, in main
2024-08-10T12:42:28.765946893Z     execute_from_command_line(sys.argv)
2024-08-10T12:42:28.765989494Z   File "/usr/local/lib/python3.10/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
2024-08-10T12:42:28.766167802Z     utility.execute()
2024-08-10T12:42:28.766184102Z   File "/usr/local/lib/python3.10/site-packages/django/core/management/__init__.py", line 436, in execute
2024-08-10T12:42:28.766399611Z     self.fetch_command(subcommand).run_from_argv(self.argv)
2024-08-10T12:42:28.766412212Z   File "/usr/local/lib/python3.10/site-packages/django/core/management/base.py", line 413, in run_from_argv
2024-08-10T12:42:28.766581819Z     self.execute(*args, **cmd_options)
2024-08-10T12:42:28.766609720Z   File "/usr/local/lib/python3.10/site-packages/django/core/management/base.py", line 459, in execute
2024-08-10T12:42:28.766668022Z     output = self.handle(*args, **options)
2024-08-10T12:42:28.766681023Z   File "/code/communications/management/commands/update_emotions.py", line 15, in handle
2024-08-10T12:42:28.773220293Z     for item in SkypeUtils.static["items"]:
2024-08-10T12:42:28.773247795Z   File "/usr/local/lib/python3.10/site-packages/skpy/util.py", line 91, in __get__
2024-08-10T12:42:28.773361899Z     return self.fget.__get__(None, owner)()
2024-08-10T12:42:28.773374900Z   File "/usr/local/lib/python3.10/site-packages/skpy/util.py", line 227, in wrapper
2024-08-10T12:42:28.773394301Z     cache[key] = fn(*args, **kwargs)
2024-08-10T12:42:28.773400501Z   File "/usr/local/lib/python3.10/site-packages/skpy/util.py", line 277, in static
2024-08-10T12:42:28.773655111Z     return SkypeConnection.externalCall("GET", json.get("pes_config")).json()
2024-08-10T12:42:28.773665412Z   File "/usr/local/lib/python3.10/site-packages/skpy/conn.py", line 123, in externalCall
2024-08-10T12:42:28.773669612Z     raise SkypeApiException("{0} response from {1} {2}".format(resp.status_code, method, url), resp)
2024-08-10T12:42:28.773810018Z skpy.core.SkypeApiException: ('409 response from GET https://static-asm.secure.skypeassets.com/pes/v1/configs/{some_id_here}/views/default', <Response [409]>)
```

### Explain your code

call static

### SkPy version

0.10.7

### Python version

3.10

**Comment by Terrance** [comment]
It looks like Skype for Web doesn't make that call any more -- I'm guessing they've transitioned to a list of emoticons built into the JavaScript code instead.  There does seem to be an endpoint to look up individual emoticons (e.g. [here for `(hi)`](https://static-asm.secure.skypeassets.com/pes/v1/emoticons/hi/views/meta.en-GB)), but I can't see a way to get a list of them any more.

**Comment by mihalt** [comment]
> It looks like Skype for Web doesn't make that call any more -- I'm guessing they've transitioned to a list of emoticons built into the JavaScript code instead. There does seem to be an endpoint to look up individual emoticons (e.g. [here for `(hi)`](https://static-asm.secure.skypeassets.com/pes/v1/emoticons/hi/views/meta.en-GB)), but I can't see a way to get a list of them any more.

oh, not good. So, do I understand correct that the only way to get full old emotions ids is to fetch saved from my old database and get each of them separately? 

Do you have this ids list? Or should I send you?

**Comment by Terrance** [comment]
Note that the list of emoticons SkPy fetched was an old (and likely incomplete) list in any case, probably due to requesting against a fake version `0.0.0.0`.  If you have a cached copy of the static data then you can continue to use it for your own lookups -- I don't think I want to ship a copy with SkPy though as everything else is currently retrieved directly from Skype servers.

There is a [support page](https://support.microsoft.com/en-us/skype/what-is-the-full-list-of-emoticons-01af0c65-529f-4a4d-8e3a-a393033a359a) with the "full list" of all emoticons, which looks like it might actually live up to its name (it includes all the hidden ones too), so you could probably parse those tables into an updated list, though the IDs aren't there.  In most if not all cases, the ID should be one of the shortcuts, but when there are multiple you'll have to figure out which one it is.

**Comment by mihalt** [comment]
> There is a [support page](https://support.microsoft.com/en-us/skype/what-is-the-full-list-of-emoticons-01af0c65-529f-4a4d-8e3a-a393033a359a) with the "full list" of all emoticons, which looks like it might actually live up to its name (it includes all the hidden ones too), so you could probably parse those tables into an updated list, though the IDs aren't there. In most if not all cases, the ID should be one of the shortcuts, but when there are multiple you'll have to figure out which one it is.

Yes, not all. For example "Grinning face with big eyes". Its id "grinningfacewithbigeyes" doesn't match with skype url. It would be really nice to have any built-in possibility of getting fresh emotions. Or at least to be able to use previous old with mock or something like that. 

Can you also maybe use any built-in api to use previously mentioned https://static-asm.secure.skypeassets.com/pes/v1/emoticons/hi/views/meta.en-GB too?

**Comment by Terrance** [comment]
https://static-asm.secure.skypeassets.com/pes/v1/configs/4b4a007d331244a1ab2e45457708c8b7/views/default 

Has the static endpoint just started working again?  Requesting config gives a static URL of  which seems to be working as before.  A temporary issue perhaps?

**Comment by mihalt** [comment]
Indeed, it woks now, ahaha. It's a luck when problem is resolving by itself. Thank's for your attention!

**Comment by mihalt** [comment]
@Terrance 

Do you know, that another strange problem is now? When I try to get
`skpy_emotions = [(item['id'], item['shortcuts'], item['visible']) for item in SkypeUtils.static["items"]]`

I get something like this 
```
requests.exceptions.SSLError: HTTPSConnectionPool(host='static-asm.secure.skypeassets.com', port=443): Max retries exceeded with url: /pes/v1/configs/4b4a007d331244a1ab2e45457708c8b7/views/default (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1007)')))
```

Also you can check by yourself this example https://static-asm.secure.skypeassets.com/pes/v1/emoticons/gingerkeepfit/views/default_100_f

It works from Windows browser. But I can't get this url from Linux systems while I not do something of this:
```
url = "https://static-asm.secure.skypeassets.com/pes/v1/emoticons/gingerkeepfit/views/default_100_f"
request.get(url, verify=False)
# or
requests.get(url, verify='/code/communications/static-asm-secure-skypeassets-com-chain.pem') # It was downloaded manually from brouser
```

I hope that problem will be resolved by itself, but maybe you should do something to fix it now.

**Comment by Terrance** [comment]
That's a problem with your environment, not with SkPy -- presumably the certificate trust store you're using is out of date.  Compare with a plain call to `requests.get()`, or using e.g. `curl` outside of Python.

**Comment by mihalt** [comment]
> That's a problem with your environment, not with SkPy -- presumably the certificate trust store you're using is out of date. Compare with a plain call to `requests.get()`, or using e.g. `curl` outside of Python.

 But earlier it worked. Yes, that's true that requests.get() and curl doesn't work too on linux for this url. I gave you example of pure `request.get(url, verify=False)` that works. I think a lot of users use linux system with your library. And it is for my local wsl and for production linux both. So, I don't think that the problem is just in OS. Or maybe if it is, can you recommend something to prevent an error? I tried a lot of different ways, but it doesn't work. 
```
apt-get install --reinstall ca-certificates
pip install --upgrade certifi
update-ca-certificates

# Commands below just return 1 certificate, but I need a chain.
curl -kw %{certs} https://static-asm.secure.skypeassets.com >> /usr/local/lib/python3.10/site-packages/certifi/cacert.pem
openssl s_client -verify 5 -connect static-asm.secure.skypeassets.com:443 -showcerts >> /usr/local/lib/python3.10/site-packages/certifi/cacert.pem
```

**Comment by Terrance** [comment]
Please don't edit your comments to add mentions, repeatedly pinging me on replies is not going to get you a response any faster.

---

As for what I'd started digging into at the time, it looks like [an intermediate SSL certificate isn't being provided by the server](https://www.ssllabs.com/ssltest/analyze.html?d=static%2dasm.secure.skypeassets.com&s=96.16.55.134), hence you only seeing one certificate,  They're presumably expecting you to obtain it by following [AIA information](https://datatracker.ietf.org/doc/html/rfc5280.html#section-4.2.2.1), a certificate extension handled by browsers but not much else.

> You can just correct it in your lib.

No, this is not SkPy's problem to correct.  SkPy defers to Requests to handle its HTTP communications, which in turn relies on the underlying transports of urllib/Python and the system/fallback certificate store to verify TLS certificates from each server.

Setting `verify=False` effectively disables this security mechanism, meaning a malicious actor can spoof requests on any domain with their own self-signed certificates.  Baking in a custom certificate store with the intermediates would mean ensuring coverage across all possible domains that might be contacted (many of which are regional and therefore not easily discoverable), and would require updates any time Microsoft decides to start using a different certificate issued by a new CA.  Implementing AIA appears to be non-trivial, which is probably why [neither Python](https://github.com/python/cpython/issues/62817) [nor curl](https://github.com/curl/curl/blob/56998277f0571d1b01939e6097068835db09f765/docs/TODO#L873-L882) support it currently.  Correcting TLS problems from the server also feels like the wrong side of the line of things to try and work around.

If you don't care about the risks, you can adjust the certificate verification for your own code via the [exposed Requests session](https://skpy.t.allofti.me/reference/internal.html#skpy.conn.SkypeConnection.sess):

```python
sk = Skype(...)
sk.conn.sess.verify = False  # or path to your own bundle of CAs
```
